stores/metrics.go

```diff
@@ -82,42 +82,49 @@ func (dbContractSetMetric) TableName() string      { return "contract_sets" }
 func (dbContractSetChurnMetric) TableName() string { return "contract_sets_churn" }
 func (dbPerformanceMetric) TableName() string      { return "performance" }
 
-func scopeTimeRange(tx *gorm.DB, after, before time.Time) *gorm.DB {
-	if after != (time.Time{}) {
-		tx = tx.Where("timestamp > ?", unixTimeMS(after))
-	}
-	if before != (time.Time{}) {
-		tx = tx.Where("timestamp <= ?", unixTimeMS(before))
-	}
-	return tx
+// scopePeriods is the core of all methods retrieving metrics. By using integer
+// division rounding combined with a GROUP BY operation, all rows of a table are
+// split into intervals and the row with the lowest timestamp for each interval
+// is returned. The result is then joined with the original table to retrieve
+// only the metrics we want.
+func (s *SQLStore) findPeriods(tx *gorm.DB, dst interface{}, table string, start time.Time, n uint64, interval time.Duration) error {
+	end := start.Add(time.Duration(n) * interval)
+	// inner groups all metrics within the requested time range into periods of
+	// 'interval' length and gives us the min timestamp of each period.
+	inner := tx.Model(dst).
+		Select("MIN(timestamp) AS min_time, (timestamp - ?) / ? * ? AS period", unixTimeMS(start), interval.Milliseconds(), interval.Milliseconds()).
```

Comment:
is this safe `(timestamp - ?) / ? * ? AS period` - don't you have to `CEIL` to get the closest multiple of?